syntax = "proto3";

package ingestion;

option go_package = ".;proto_gen";

// InternalIngestionService provides an API for the main backend to read
// spans from the BadgerDB WAL.
service InternalIngestionService {
  // ReadSpans reads a batch of spans from the WAL, starting after the
  // specified ULID. This is a server-streaming RPC.
  rpc ReadSpans(ReadSpansRequest) returns (stream ReadSpansResponse) {}
}

// ReadSpansRequest defines the parameters for requesting a batch of spans.
message ReadSpansRequest {
  // The ULID of the last span successfully processed by the client.
  // The server will return spans that were received *after* this key.
  // If empty, the stream will start from the oldest available span.
  bytes start_key_ulid = 1;

  // The maximum number of spans to return in the batch.
  uint32 batch_size = 2;
}

// ReadSpansResponse contains a single span from the WAL.
message ReadSpansResponse {
  // The ULID key of the span, which also serves as its timestamp.
  bytes key_ulid = 1;

  // The raw, serialized protobuf bytes of the OTel span.
  bytes span_bytes = 2;

  // The raw, serialized protobuf bytes of the OTel resource.
  bytes resource_bytes = 3;
}
